wicker/l5ml_datastore/v2/index_ops.py

```diff
@@ -0,0 +1,253 @@
+"""Collection of functions to manipulate datasets.
+"""
+import copy
+from typing import Optional, Tuple, TypeVar
+
+import pyarrow  # type: ignore
+import pyarrow.compute  # type: ignore
+
+from wicker.l5ml_datastore import cpp_extensions
+from wicker.core.datasets import AbstractDataset
+from wicker.core.definitions import Example
+from wicker.core.errors import WickerDatastoreException, WickerSchemaException
+from wicker.schema.schema import ArrayField, DatasetSchema
+
+CellSpecification = cpp_extensions.CellSpecification
+ColumnHistorySpecification = cpp_extensions.ColumnHistorySpecification
+WindowSpecification = cpp_extensions.WindowSpecification
+
+T = TypeVar("T")
+
+
+def _assert_not_none(x: Optional[T]) -> T:
+    assert x is not None
+    return x
+
+
+def merge_schemas(
+    left: DatasetSchema, right: DatasetSchema, left_required: bool = True, right_required: bool = True
+) -> DatasetSchema:
+    """Create a new schema by combining left and right schemas. Raises WickerSchemaException if the two
+    schemas are incompatible.
+    The left_required and right_required flags control which fields are required in the merged schema.
+    :param left: One of the schemas to merge.
+    :param right: The other schema to merge.
+    :param left_required: If a field comes from the left schema, it is required in the final schema iff it is required
+      in the left schema and left_required is set.
+    :param right_required: If a field comes from the right schema, it is required in the final schema iff it is required
+      in the right schema and right_required is set.
+    :return: A new schema with fields from both left and right.
+    """
+    if left.primary_keys != right.primary_keys:
+        raise WickerSchemaException(
+            "The schemas have different primary keys: Got " f"{left.primary_keys} and  {right.primary_keys}"
+        )
+    # Make deep copies of the schemas so that we can change the "required" status.
+    left = copy.deepcopy(left)
+    right = copy.deepcopy(right)
+
+    new_fields = left._columns
+    if not left_required:
+        for name, field in new_fields.items():
+            field.required = name in left.primary_keys
+    for name, field in right._columns.items():
+        if name in new_fields:
+            if not new_fields[name].is_compatible(field):
+                raise WickerSchemaException(
+                    f"Can not merge schemas. Field '{name}' seen with two different types. "
+                    f"{type(new_fields[name])}  {type(field)}"
+                )
+            if right_required and field.required:
+                new_fields[name].required = True
+        else:
+            field.required = field.required and right_required
+            new_fields[name] = field
+    return DatasetSchema(fields=list(new_fields.values()), primary_keys=left.primary_keys)
+
+
+def join_datasets(
+    left: AbstractDataset[Example], right: AbstractDataset[Example], how: str = "outer"
```

Comment:
It would be great if how could be typed as a `Literal['left', 'right', 'outer', 'inner']`